datafusion/sqllogictest/test_files/coalesce.slt

```diff
@@ -35,7 +35,7 @@ select coalesce(null, null);
 NULL
 
 # cast to float
-query RT
+query IT
```

Comment:
likewise this looks reasonable (coerce has changed to have type `Int64` because the first argument is Int64. But why did the report of `arrow_typeof` not change 🤔

Reply:
I also notice that, it seems types are computed before the actual result, so the type is not coerced

https://github.com/apache/datafusion/blob/6480020e695ebbe2b81e8971c3ee0e9e7ec124b0/datafusion/sqllogictest/src/engines/datafusion_engine/runner.rs#L73-L80

And, it is not a trivial fix, I tried to get the return type from batches, but failed on aggregate cases where the expected rows are different from the batches (input rows). 

Do you think we need to fix it prior to this PR?
Given that we have arrow_typeof, I don't think it is necessary, but it is somehow confusing if we read carefully about the type besides `query`.